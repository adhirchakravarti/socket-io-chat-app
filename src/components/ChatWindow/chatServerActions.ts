import {Message} from "./types";
import {
    SEND_MESSAGE,
    SET_USERNAME,
    RECEIVE_MESSAGE,
    GET_INITIAL_USERNAME
} from "./chatServerConstants";

interface sendMessageAction {
    type: typeof SEND_MESSAGE;
    payload: {
        message: Message;
    };
}

interface receiveMessageAction {
    type: typeof RECEIVE_MESSAGE;
    payload: {
        message: Message;
    };
}

interface setUsernameAction {
    type: typeof SET_USERNAME;
    payload: {
        userName: string;
    };
}

interface getInitialUserNameAction {
    type: typeof GET_INITIAL_USERNAME;
}

export function sendMessage(message: Message): sendMessageAction {
    return {
        type: SEND_MESSAGE,
        payload: {
            message
        }
    };
}

export function receiveMessage(message: Message): receiveMessageAction {
    return {
        type: RECEIVE_MESSAGE,
        payload: {
            message
        }
    };
}

export function setUsername(userName: string): setUsernameAction {
    return {
        type: SET_USERNAME,
        payload: {
            userName
        }
    };
}

export function getInitialUserName(): getInitialUserNameAction {
    return {
        type: GET_INITIAL_USERNAME
    };
}
