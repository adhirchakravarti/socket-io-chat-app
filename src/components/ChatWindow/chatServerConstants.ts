export const WELCOME = "chatClient/WELCOME";
export const GET_INITIAL_USERNAME = "chatServer/GET_INITIAL_USERNAME";
export const SET_USERNAME = "chatServer/SET_USERNAME";
export const SET_USERNAME_SUCCESS = "chatClient/SET_USERNAME_SUCCESS";
export const SEND_MESSAGE = "chatServer/SEND_MESSAGE";
export const RECEIVE_MESSAGE = "chatClient/RECEIVE_MESSAGE";
export const DISCONNECT = "chatServe/DISCONNECT";
