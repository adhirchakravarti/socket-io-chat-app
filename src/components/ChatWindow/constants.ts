export const CREATE_CONNECTION = "app/Chat/CREATE_CONNECTION";
export const SET_USERNAME = "app/Chat/SET_USERNAME";
export const CLOSE_CONNECTION = "app/Chat/CLOSE_CONNECTION";
export const SEND_MESSAGE = "app/Chat/SEND_MESSAGE";
export const RECEIVE_MESSAGE = "app/Chat/RECEIVE_MESSAGE";
export const SERVER_URL = "http://localhost:3002";
