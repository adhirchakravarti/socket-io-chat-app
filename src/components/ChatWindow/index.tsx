/* eslint-disable @typescript-eslint/no-unused-vars */
import React from "react";
import {connect} from "react-redux";
import {compose, Dispatch} from "redux";
import Container from "@material-ui/core/Container";
import Paper from "@material-ui/core/Paper";
import {v4 as uuidv4} from "uuid";
// import Grid from "@material-ui/core/Grid";
import Box from "@material-ui/core/Box";
import {withStyles, Theme, StyleRules} from "@material-ui/core/styles";
import ChatInput from "../ChatInput";
import ChatMessageList from "../ChatMessageList";
import {Message} from "../../types/index";
import {
    sendMessage,
    receiveMessage,
    getInitialUserName,
    resetUnreadMessageCount
} from "../../store/chatServerActions";

const styles = (theme: Theme): StyleRules => ({
    root: {
        width: "100%",
        display: "flex",
        flexDirection: "column",
        flexWrap: "nowrap",
        backgroundColor: theme.palette.background.paper
    },
    chatContainer: {
        display: "flex",
        flexDirection: "column",
        flexGrow: 1
    },
    listContainer: {
        display: "flex",
        height: "67vh",
        [theme.breakpoints.down("md")]: {
            height: "70vh"
        },
        [theme.breakpoints.down("xs")]: {
            height: "72vh"
        }
    },
    input: {
        display: "flex",
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "stretch"
    }
});

type ChatWindowClasses = {
    root: string;
    chatContainer: string;
    listContainer: string;
    input: string;
};

type ComponentState = {
    socket: SocketIO.Socket;
    messages: Message[];
    userName: string;
};

interface ChatWindowProps {
    classes: ChatWindowClasses;
    messages: Message[];
    userName: string;
    socketId: string;
    clock: string;
    sendMessageOnCtrlEnter: string;
    getInitialUserName: () => Dispatch;
    sendNewMessage: (message) => Dispatch;
    receiveNewMessage: (message) => Dispatch;
    resetUnreadMessageCountAction: () => Dispatch;
}

const mapStateToProps = (state) => {
    console.log("reducer state = ", state);
    return {
        messages: state.chatReducer.messages,
        userName: state.chatReducer.userName,
        socketId: state.chatReducer.socketId,
        clock: state.chatReducer.clock,
        sendMessageOnCtrlEnter: state.chatReducer.sendMessageOnCtrlEnter
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        getInitialUserName: () => dispatch(getInitialUserName()),
        sendNewMessage: (message: Message) => dispatch(sendMessage(message)),
        receiveNewMessage: (message: Message) => dispatch(receiveMessage(message)),
        resetUnreadMessageCountAction: () => dispatch(resetUnreadMessageCount())
    };
};

class ChatWindow extends React.Component<ChatWindowProps> {
    state = {
        messages: [],
        userName: ""
    };

    componentDidMount() {
        this.props.resetUnreadMessageCountAction();
    }

    handleInputMessage = (message: string): void => {
        const newMessage = {
            id: uuidv4(),
            text: message,
            sender: this.props.userName,
            socketId: this.props.socketId,
            time: new Date().getTime()
        };
        const {sendNewMessage} = this.props;
        sendNewMessage(newMessage);
    };
    render() {
        const {classes} = this.props;
        return (
            <Container maxWidth="xl" className={classes.root} disableGutters>
                <Paper className={classes.chatContainer} elevation={3}>
                    <Box component="div" className={classes.listContainer}>
                        <ChatMessageList
                            messages={this.props.messages}
                            socketId={this.props.socketId}
                            clock={this.props.clock}
                        />
                    </Box>
                    <Box className={classes.input}>
                        <ChatInput
                            onSubmit={this.handleInputMessage}
                            sendMessageOnCtrlEnter={this.props.sendMessageOnCtrlEnter}
                        />
                    </Box>
                </Paper>
            </Container>
        );
    }
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(withStyles(styles)(ChatWindow));
