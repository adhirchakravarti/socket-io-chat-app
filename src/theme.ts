import {createMuiTheme} from "@material-ui/core";

export const lightTheme = createMuiTheme({
    palette: {
        type: "light",
        primary: {
            main: "#0d47a1"
        },
        secondary: {
            main: "#689f38"
        }
    }
});

export const darkTheme = createMuiTheme({
    palette: {
        type: "dark",
        primary: {
            main: "#29b6f6"
        },
        secondary: {
            main: "#689f38"
        },
        background: {
            paper: "#212121",
            default: "#000"
        },
        contrastThreshold: 5
    }
});
