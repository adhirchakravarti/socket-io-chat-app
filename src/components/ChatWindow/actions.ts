import {CREATE_CONNECTION, RECEIVE_MESSAGE} from "./constants";
import {ChatActionTypes, Message} from "./types";

export function createConnection(): ChatActionTypes {
    return {
        type: CREATE_CONNECTION
    };
}

export function receiveMessage(message: Message): ChatActionTypes {
    return {
        type: RECEIVE_MESSAGE,
        payload: {
            message
        }
    };
}
