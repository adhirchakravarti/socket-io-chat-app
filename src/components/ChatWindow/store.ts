import {createStore, combineReducers, applyMiddleware} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import createSocketIoMiddleware from "redux-socket.io";
import {SERVER_URL} from "./constants";
import chatReducer from "./reducer";
import io from "socket.io-client";

const socket = io(SERVER_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, "chatServer/");
const middleWares = [socketIoMiddleware];
// import {initialState} from "./reducer";

// function reducer(state = {}, action) {
//     switch (action.type) {
//         case "message":
//             return Object.assign({}, {message: action.data});
//         default:
//             return state;
//     }
// }

const rootReducer = combineReducers({
    chatReducer
});
const enhancer = composeWithDevTools(applyMiddleware(...middleWares));

const store = createStore(rootReducer, enhancer);

store.subscribe(() => {
    console.log("new client state", store.getState());
});
// store.dispatch({type: "chatServer/hello", data: "Hello!"});
setTimeout(() => {
    store.dispatch({type: "chatServer/SET_USERNAME", payload: {userName: "Jack"}});
}, 20000);

export default store;
