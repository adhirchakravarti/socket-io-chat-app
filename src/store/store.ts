import {createStore, combineReducers, applyMiddleware} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import createSocketIoMiddleware from "redux-socket.io";
import {SERVER_URL} from "./constants";
import chatReducer from "./reducer";
import io from "socket.io-client";

const socket = io(SERVER_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, "chatServer/");
const middleWares = [socketIoMiddleware];

const rootReducer = combineReducers({
    chatReducer
});
const enhancer = composeWithDevTools(applyMiddleware(...middleWares));

const store = createStore(rootReducer, enhancer);

const savedSettings = JSON.parse(sessionStorage.getItem("settings"));
if (savedSettings && Object.keys(savedSettings).length > 0) {
    const savedUserName = savedSettings.userName;
    store.dispatch({type: "chatClient/LOAD_SETTINGS"});
    setTimeout(
        () => store.dispatch({type: "chatServer/SET_USERNAME", payload: {userName: savedUserName}}),
        1000
    );
}

export default store;
