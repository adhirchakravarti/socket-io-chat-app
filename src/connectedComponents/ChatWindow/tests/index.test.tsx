import React from "react";
import {render} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {ThemeProvider} from "@material-ui/core/styles";
import {StylesProvider} from "@material-ui/styles/";
import CssBaseline from "@material-ui/core/CssBaseline";

import GlobalResponsiveFont from "../../../globalResponsiveFont";
import {lightTheme} from "../../../theme";
import store from "../../../store/store";

import ChatWindow from "../index";

const mockProps = {
    maxHeight: 73,
    parentHeight: 600,
    isPortrait: true
};

function getPreparedPage(store, theme): JSX.Element {
    const generateClassName = (rule, styleSheet) =>
        `${styleSheet.options.classNamePrefix}-${rule.key}`;
    return (
        <Provider store={store}>
            <StylesProvider generateClassName={generateClassName}>
                <ThemeProvider theme={theme}>
                    <CssBaseline />
                    <GlobalResponsiveFont />
                    <MemoryRouter>
                        <ChatWindow {...mockProps} />
                    </MemoryRouter>
                </ThemeProvider>
            </StylesProvider>
        </Provider>
    );
}

describe("<ChatWindow />", () => {
    it("renders correctly and matches snapshot", () => {
        const {asFragment} = render(getPreparedPage(store, lightTheme));
        const initialRender = asFragment();
        expect(initialRender).toMatchSnapshot();
    });
    it("does not log errors in the console", () => {
        const spy = jest.spyOn(global.console, "error");
        render(getPreparedPage(store, lightTheme));
        expect(spy).not.toHaveBeenCalled();
    });
});
