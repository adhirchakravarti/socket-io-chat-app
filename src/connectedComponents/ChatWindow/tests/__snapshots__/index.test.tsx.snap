// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatWindow /> renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root makeStyles-chatContainer makeStyles-chatContainer MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      class="MuiBox-root MuiBox-root makeStyles-listContainer makeStyles-listContainer"
    >
      <ul
        class="MuiList-root makeStyles-list MuiList-padding"
      />
    </div>
    <div
      class="MuiBox-root MuiBox-root makeStyles-input"
    >
      <div
        class="MuiBox-root MuiBox-root makeStyles-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-input MuiFormControl-marginDense"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="true"
          >
            Chat
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              placeholder="Enter chat message and click the send button"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
              >
                <span>
                  Chat
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Send
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
